src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php

```diff
@@ -58,8 +58,18 @@ public function getTypeExtensions($name)
             foreach ($this->typeExtensionServices[$name] as $serviceId => $extension) {
                 $extensions[] = $extension;
 
+                if (method_exists($extension, 'getExtendedTypes')) {
```

Comment:
This validation prevents to use different instances of the same class to be used for different extensions. So it prevents the only advantage of `extended_type` attribute that I mentioned in https://github.com/symfony/symfony/pull/24530#issuecomment-361418781 as people will be forced to implement the method in 5.0 when its part of the interface. So they need to return an empty array and then overwrite it with the service attribute.

IMO we do not need the validation anymore as now both the registration and implementation uses getExtendedTypes. So there should not be a mismatch anymore unless you wire something wrong manually.

Reply:
```php
class FooBarTypeExtension extends AbstractTypeExtension
{
    public function __construct(string $type)
    {
    }

    public static function getExtendedTypes(): iterable
    {
        return array(
            FooType::class,
            BarType::class,
        );
    }
}
```

```yaml
App\Form\FooBarTypeExtension:
    arguments: ['foo']
    tags:
        - name: form.type_extension
          extended_type: FooType

App\Form\FooBarTypeExtension:
    arguments: ['bar']
    tags:
        - name: form.type_extension
          extended_type: BarType
```

Do I understand correctly that you have something like this in mind?

Reply:
Yes. By returning both types from getExtendedTypes, it actually works even with the validation. So I'm fine keeping it like this. But I think it would make more sense to remove the validation as it should not matter what is returned by getExtendedTypes if you set the extended_type attribute (and you might not even know that in advance if you create a dynamic form extension).

Reply:
I wonder if we shouldn't rather deprecate the `extended_type` attribute too instead I could imagine that using it adds more confusion than an implementation where `getExtendedTypes()` is the single source of truth.

Reply:
I agree that it potentially is complicating things for no real use-case. But I also like that it allows to configure extensions dynamically via DI (instead of hardcoding it into a class). As I said in https://github.com/symfony/symfony/pull/24530#issuecomment-361418781, its similar to the command attribute of the console.command tag. So it would be consistent to keep it. But I would not object to deprecate it as well.

Reply:
Let's keep it for now. We can still consider to deprecate it later when we noticed that it adds too much confusion.